article/_posts/2019-09-26-VQ-VAE.md

```diff
@@ -0,0 +1,163 @@
+output $$z_{e}(x)$$. The discrete latent variables $$z$$ are then calculated by
+a nearest neighbor look-up using the shared embedding space $$e$$:
+
+![](/article/images/VQ-VAE/Latent_variable.png)
+
+This forward computation pipeline can be seen as a regular autoencoder with a
+particular non-linearity that maps the latents to 1-of-$$K$$ embedding vectors.
+
+The posterior and prior distributions are categorical, and the samples drawn
+from these distributions index an embedding table. Thus:
+
+1. Encoders model a categorical distribution, sampling from which integral
+values are obtained
+2. These integral values are used to index a dictionary of embeddings.
+3. The indexed values are then passed on to the decoder.
+
+During forward computation the nearest embedding $$z_{q}(x)$$ is passed to the
+decoder.
+
+However, the mapping from $$z_{e}$$ to $$z_{q}$$ is not differentiable. In order
+to alleviate this, authors use a *straight-through estimator*, meaning that the
+gradient $$\nabla_{z}L$$ from the decoder input $$z_{q}(x)$$ (quantized) are
+directly copied to the encoder output $$z_{e}(x)$$ (continuous) during the
+backward pass.
+
+Therefore, in order to learn the embedding space, authors use a dictionary
+learning technique, Vector Quantisation (VQ). The VQ objective uses the *l2*
+error to move the make the latent code closer to the continuous vector it was
+matched to.
+
+Thus, the loss function has three components:
+
+![](/article/images/VQ-VAE/Loss_function.png)
+
+- The first term is the reconstruction loss (which optimizes the decoder and the
+encoder). Since a uniform prior for $$z$$ is assumed, the KL term that usually
+appears in the ELBO is constant w.r.t. the encoder parameters and can thus be
+ignored for training.
+- The second term is the contribution of the VQ, which optimises the embeddings.
+*sg* stands for "stop gradient" and is defined as identity at forward
+computation time and has zero partial derivatives, constraining its operand to
+be a non-updated constant.
+- The third term is a regularization term they name as *committment loss* which
+controls the volume of the latent space by forcing the encoder to *commit* to
+the latent code it matched with, and avoid its output space growing unboundedly.
+
+The second contribution of this work consists in learning the prior
+distribution. During the training phase, the prior $$p(z)$$ is a uniform
+categorical distribution. After the training is done, they fit an
+"autoregressive distribution" over the space of latent codes.
+
+
+# Results
+
+## Images
+
+Three complex image datasets (CIFAR10, ImageNet, and DeepMind Lab) were used.
+
+The proposed model is compared to the standard continuous VAE framework. It
+seems to achieve similar sample quality, while taking advantage of the discrete
+latent space. The model still performs well when using a PixelCNN decoder on top
+of the VQ-VAE.
+
+![](/article/images/VQ-VAE/Results_images.png)
```

Comment:
You could put a higher resolution image, so we could see that the reconstruction is somewhat blurry.
Also, you could add the figure 3, where we see samples that do not correspond to training images.

Reply:
I tend to reduce the original size of the screenshots to avoid overly large images and save storage space. I grant that at times this means that the quality is not the ideal at times. But we can visit the original paper in that case.

Added figure 3 in commit 8eb4c45 .